torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -1,257 +1,293 @@
+  bool inlineIfTooSmall(Node* n) {
+    JIT_ASSERT(n->kind() == prim::DifferentiableGraph);
+    auto& subgraph = getSubgraph(n);
+    size_t i = 0;
+    for (const auto node __attribute__((unused)) : subgraph.nodes()) {
+      if (++i >= minSubgraphSize_) {
+        return false;
+      }
+    }
+
+    unmergeGroup(n);
     return true;
   }
-  JIT_ASSERT(v->data.size() == 1);
-  auto * node = *v->data.begin();
-  if (node->kind() == prim::Constant) {
-    return false;
+
+  // Move nodes from producerGroup's subgraph to the top-level graph.
+  // This destroys `producerGroup`.
+  std::vector<Node*> unmergeGroup(Node* producerGroup) {
+    JIT_ASSERT(producerGroup->kind() == prim::DifferentiableGraph);
+
+    std::vector<Node*> temporary_nodes;
+    auto& producerSubgraph = getSubgraph(producerGroup);
+
+    // Initialize a map of inner graph values to outer graph values
+    std::unordered_map<const Value*, Value*> innerToOuter;
+    const auto innerInputs = producerSubgraph.inputs();
+    const auto outerInputs = producerGroup->inputs();
+    for (size_t i = 0; i < innerInputs.size(); ++i) {
+      innerToOuter[innerInputs[i]] = outerInputs[i];
+    }
+
+    // Clone all nodes
+    for (auto inner : producerSubgraph.nodes()) {
+      Node* outer = block_->owningGraph()->createClone(
+          inner, [&](Value* k) -> Value* { return innerToOuter.at(k); });
+      outer->insertBefore(producerGroup);
+      temporary_nodes.emplace_back(outer);
+      const auto innerOutputs = inner->outputs();
+      const auto outerOutputs = outer->outputs();
+      for (size_t i = 0; i < innerOutputs.size(); ++i)
+        innerToOuter[innerOutputs[i]] = outerOutputs[i];
+    }
+
+    // Replace uses of producerGroup outputs and destroy the producer
+    const auto subgraphOutputs = producerSubgraph.outputs();
+    for (size_t i = 0; i < subgraphOutputs.size(); ++i) {
+      const auto outerOutput = innerToOuter.at(subgraphOutputs[i]);
+      producerGroup->outputs()[i]->replaceAllUsesWith(outerOutput);
+    }
+    producerGroup->destroy();
+
+    return temporary_nodes;
   }
-  return isDifferentiable(node);
-}
 
-static detail::DynamicDAG<Node*> make_dependency_graph(Block * block) {
-  detail::DynamicDAG<Node*> dag;
-  std::unordered_map<Node*,detail::Vertex<Node*>*> node_to_vertex;
-  // NB: the block's param and return nodes are not in the dependency graph.
-  for (Node * node : block->nodes()) {
-    node_to_vertex[node] = dag.newVertex(node);
+  // Combine the nodes in two groups together. The nodes will end up in
+  // `consumerGroup`, and `producerGroup` will be deleted.
+  void mergeGroups(Node* consumerGroup, Node* producerGroup) {
+    // Extract the nodes in `producerGroup` into the outer graph
+    const auto nodes = unmergeGroup(producerGroup);
+    // Then merge them into `consumerGroup`
+    for (auto it = nodes.rbegin(); it != nodes.rend(); ++it) {
+      mergeNodeIntoGroup(consumerGroup, *it);
```

Comment:
This seems complicated. I'm assuming the two groups are adjacent to each other in the topological indexing (is this correct?). If so, couldn't we keep the inputs of producer group, the outputs of consumer group, and then smash the nodes of the groups together instead of unpacking the group and then moving each node individually?

Reply:
We'd also need to add the inputs of the consumer to the producer, if and only if the inputs of the consumer are not generated by the producer (and the inverse for outputs). This ends up basically being equivalent to the code we have here in complicatedness.

Since we need "unmergeGroup" anyway for inlining subgraphs that are too small, I think it's better to just re-use it here.